packages/talos_messenger_actions/src/kafka/producer.rs

```diff
@@ -0,0 +1,143 @@
+        match result {
+            Ok(msg) => {
+                info!("Message {:?} {:?}", msg.key(), msg.offset());
+                // TODO: GK - what to do on error? Panic?
+                let _ = block_on(self.tx_feedback_channel.send(MessengerChannelFeedback::Success(
+                    version,
+                    "kafka".to_string(),
+                    delivery_opaque.total_publish_count,
+                )));
+            }
+            Err(err) => {
+                // TODO: GK - what to do on error? Panic?
+                let _ = block_on(self.tx_feedback_channel.send(MessengerChannelFeedback::Error(version, err.0.to_string())));
+            }
+        }
+    }
+}
+
+// Kafka Producer
+// #[derive(Clone)]
+pub struct KafkaProducer<C: ProducerContext + 'static = DefaultProducerContext> {
+    producer: ThreadedProducer<C>,
+    topic: String,
+}
+
+impl KafkaProducer {
+    pub fn new(config: &KafkaConfig) -> Self {
+        let client_config = config.build_producer_config();
+        let producer = ThreadedProducer::from_config(&client_config).expect("Failed to create producer");
+        let topic = config.topic.to_owned();
+
+        Self { producer, topic }
+    }
+}
+impl<C: ProducerContext + 'static> KafkaProducer<C> {
+    pub fn with_context(config: &KafkaConfig, context: C) -> Self {
+        let client_config = config.build_producer_config();
+        let producer = ThreadedProducer::from_config_and_context(&client_config, context).expect("Failed to create producer");
+        let topic = config.topic.to_owned();
+
+        Self { producer, topic }
+    }
+
+    pub fn publish_to_topic(
+        &self,
+        topic: &str,
+        key: &str,
+        value: &str,
+        headers: Option<HashMap<String, String>>,
+        delivery_opaque: C::DeliveryOpaque,
+    ) -> Result<(), MessagePublishError> {
+        let record = BaseRecord::with_opaque_to(topic, delivery_opaque).payload(value).key(key);
+
+        let record = match headers {
+            Some(x) => record.headers(build_kafka_headers(x)),
+            None => record,
+        };
+
+        self.producer.send(record).map_err(|(kafka_error, record)| MessagePublishError {
+            reason: kafka_error.to_string(),
+            data: Some(format!(
+                "Topic={:?} partition={:?} key={:?} headers={:?} payload={:?}",
+                self.topic, record.partition, record.key, record.headers, record.payload
+            )),
+        })
+    }
+}
+
+//  Message publisher traits
+#[async_trait]
+impl MessagePublisher for KafkaProducer {
+    async fn publish_message(&self, key: &str, value: &str, headers: Option<HashMap<String, String>>) -> Result<(), SystemServiceError> {
+        let record = BaseRecord::to(&self.topic).payload(value).key(key);
+
+        let record = match headers {
+            Some(x) => record.headers(build_kafka_headers(x)),
+            None => record,
+        };
+
+        debug!("Preparing to send the Decision Message. ");
```

Comment:
Decision?

Reply:
What you're alluding to is that the code should be generic and not only deal with decisions, right?

Reply:
Yes, that should be a generic message.

I will fix that, thanks for pointing that out. 

This impl of `MessagePublisher` is actually copy from the producer I created for certifier. 

The difference between this `producer` vs the one from certifier is that, this gets more generic, allowing to have custom `delivery context` and `delivery opaque`. 

Once this going through some clean up, I would like to move this into the `certifier` producer and use it here.